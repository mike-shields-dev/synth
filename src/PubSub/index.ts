export { ControlChangeSubscriber, publishControlChange } from './ControlChange';
export type { ControlChange } from './ControlChange/types';
export { MidiNoteOnSubscriber, publishMidiNoteOn } from './MidiNoteOn';
export type { MidiNoteOn } from './MidiNoteOn/types';
export { NoteOffSubscriber, publishNoteOff } from './NoteOff';
export type { NoteOff } from './NoteOff/types';

