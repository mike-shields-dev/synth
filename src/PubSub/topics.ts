const
    MIDI = 'midi',
    MIDI_CC = 'midi.controlChange',
    MIDI_NOTE_ON = 'midi.note.on',
    MIDI_NOTE_OFF = 'midi.note.off',
    UI_CC = 'ui.controlChange',
    FOCUS = 'focus';

export {
    MIDI, 
    MIDI_CC, 
    MIDI_NOTE_ON,
    MIDI_NOTE_OFF,
    UI_CC,
    FOCUS,
}