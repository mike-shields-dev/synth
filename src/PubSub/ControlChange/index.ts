import PubSub from 'pubsub-js';
import { CONTROL } from '../topics';
import { ControlChange, ControlChangeHandler } from './types';

function publishControlChange(data: ControlChange) {
  PubSub.publish(CONTROL, data);
}

class ControlChangeSubscriber {
  private subscription: string;

  constructor(controlChangeHandler: ControlChangeHandler) {
    this.subscription = PubSub.subscribe(
      CONTROL,
      controlChangeHandler,
    );
  }

  public unsubscribe() {
    PubSub.unsubscribe(this.subscription);
  }
}

export {
  publishControlChange,
  ControlChangeSubscriber
};

