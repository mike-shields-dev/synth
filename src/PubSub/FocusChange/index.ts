import PubSub from "pubsub-js";
import { FOCUS } from '../topics';
import { FocusChangeHandler } from "./types";

function publishFocusChange(data: string) {
    PubSub.publish(FOCUS, data);
}

class FocusChangeSubscriber {
    private subscription: string;

    constructor(focusHandler: FocusChangeHandler) {
        this.subscription = PubSub.subscribe(
            FOCUS,
            focusHandler,
        );
    }

    public unsubscribe() {
        PubSub.unsubscribe(this.subscription);
    }
}

export {
    publishFocusChange,
    FocusChangeSubscriber
};
