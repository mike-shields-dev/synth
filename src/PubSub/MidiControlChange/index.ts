import PubSub from 'pubsub-js';
import { MIDI_CC } from '../topics';
import { MidiControlChange, MidiControlChangeHandler } from './types';

function publishMidiControlChange(data: MidiControlChange) {
  PubSub.publish(MIDI_CC, data);
}

class MidiControlChangeSubscriber {
  private subscription: string;

  constructor(controlChangeHandler: MidiControlChangeHandler) {
    this.subscription = PubSub.subscribe(
      MIDI_CC,
      controlChangeHandler,
    );
  }

  public unsubscribe() {
    PubSub.unsubscribe(this.subscription);
  }
}

export {
  publishMidiControlChange,
  MidiControlChangeSubscriber
};

