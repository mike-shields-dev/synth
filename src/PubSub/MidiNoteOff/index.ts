import PubSub from "pubsub-js";
import { MIDI_NOTE_OFF } from "../topics";
import { MidiNoteOff, MidiNoteOffHandler } from './types';

function publishMidiNoteOff(payload: MidiNoteOff) {
    PubSub.publish(MIDI_NOTE_OFF, payload)
}

class MidiNoteOffSubscriber {
    private subscription: string;

    constructor(noteOffHandler: MidiNoteOffHandler) {
        this.subscription = PubSub.subscribe(
            MIDI_NOTE_OFF,
            noteOffHandler,
        )
    }

    public unsubscribe() {
        PubSub.unsubscribe(this.subscription);
    };
}

export { publishMidiNoteOff, MidiNoteOffSubscriber };
