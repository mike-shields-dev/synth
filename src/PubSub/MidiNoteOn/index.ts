import PubSub from "pubsub-js";
import { MIDI_NOTE_ON } from "../topics";
import { MidiNoteOn, MidiNoteOnHandler } from "./types";

function publishMidiNoteOn(payload: MidiNoteOn) {
    PubSub.publish(MIDI_NOTE_ON, payload)
}

class MidiNoteOnSubscriber {
    private subscription: any;

    constructor(noteOnHandler: MidiNoteOnHandler) {
        this.subscription = PubSub.subscribe(
            MIDI_NOTE_ON,
            noteOnHandler,
        )
    }

    public unsubscribe() {
        PubSub.unsubscribe(this.subscription);
    };
}

export { publishMidiNoteOn, MidiNoteOnSubscriber };
