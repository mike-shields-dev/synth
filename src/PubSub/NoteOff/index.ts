import PubSub from "pubsub-js";
import { MIDI_NOTE_OFF } from "../topics";
import { NoteOff, NoteOffHandler } from './types';

function publishNoteOff(data: NoteOff) {
    PubSub.publish(MIDI_NOTE_OFF, data)
}

class NoteOffSubscriber {
    private subscription: string;

    constructor(noteOffHandler: NoteOffHandler) {
        this.subscription = PubSub.subscribe(
            MIDI_NOTE_OFF,
            noteOffHandler,
        )
    }

    public unsubscribe() {
        PubSub.unsubscribe(this.subscription);
    };
}

export { publishNoteOff, NoteOffSubscriber };
