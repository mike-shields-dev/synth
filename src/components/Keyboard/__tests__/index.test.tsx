import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { vi } from 'vitest';
import { Keyboard } from '../';
import { MIDI_NOTE_OFF, MIDI_NOTE_ON } from '../../../PubSub/topics';

describe("Keyboard", () => {
    beforeEach(() => {
        render(<Keyboard />)
    });

    it('matches snapshot', () => {
        const { container } = render(<Keyboard />);

        expect(container.firstChild).toMatchSnapshot();
    });

    it('renders 12 buttons for each key of an octave', () => {
        expect(screen.getAllByRole('button')).toHaveLength(12);
    });

    it(`each key has the correct note name for it's name attribute`, () => {
        const expectedNoteNames = ['C','D','E','F','G','A','B','Db','Eb','Gb','Ab','Bb'];
        const foundNoteNames = screen.getAllByRole('button').map((key) => key.getAttribute("name"));

        expect(expectedNoteNames).toEqual(foundNoteNames);
    });

    it('each button/key publishes to the MIDI_NOTE_ON topic when activated', async () => {
        const keys = screen.getAllByRole('button');
        const handler = vi.fn(() => null);

        PubSub.subscribe(MIDI_NOTE_ON, handler);

        for (const key of keys) {
            fireEvent.mouseDown(key)
        }

        await waitFor(() => {
            expect(handler).toHaveBeenCalledTimes(keys.length);
        });
    });

    it('each button/key publishes to the MIDI_NOTE_ON topic when deactivated', async () => {
        const keys = screen.getAllByRole('button');
        const handler = vi.fn(() => null);

        PubSub.subscribe(MIDI_NOTE_OFF, handler);

        for (const key of keys) {
            fireEvent.mouseUp(key)
            fireEvent.mouseLeave(key)
        }

        await waitFor(() => {
            expect(handler).toHaveBeenCalledTimes(keys.length * 2);
        });
    });
});