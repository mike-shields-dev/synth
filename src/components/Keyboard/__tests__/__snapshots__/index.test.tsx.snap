// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Keyboard > matches snapshot 1`] = `
<div>
  <button
    name="C"
  />
  <button
    name="D"
  />
  <button
    name="E"
  />
  <button
    name="F"
  />
  <button
    name="G"
  />
  <button
    name="A"
  />
  <button
    name="B"
  />
  <button
    name="Db"
  />
  <button
    name="Eb"
  />
  <button
    name="Gb"
  />
  <button
    name="Ab"
  />
  <button
    name="Bb"
  />
</div>
`;
